keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1121 @@
+
+This design is written to have no impact on existing cluster users for existing
+fields.
+
+A risk is that it would be possible to configure permissions which result in a
+large number of authz checks. Specifically CRD authors could craft a CRD with
+many checks required to do anything.
+
+A runtime risk is an actor with the "granular" permission could make requests
+changing many fields it doesn't have permission for, with the aim of driving
+many authz checks to overload the authz system. To mitigate this we (a) have the
+"granular" permission at all, to reduce the number of actors that could attempt
+this, and (b) we will not check all special permissions for a request once we
+know at least one has failed.
+
+## Design Details
+
+### Permission Description System
+
+Now that we have described all requirements and given the overall outline, we
+can finally state exactly how the mapping from field to permission works and how
+it is specified. Additionally, we will describe the parameterization systems. We
+will give a series of examples in {goal, built-in specification, CRD
+specification, resulting verb} tuplets.
+
+##### Everything in spec
+
+On types:
+
+```go
+// +permission-verb:"specification"
+type FooSpec struct {
+```
+
+Or on fields:
+
+```go
+  ...
+  // +permission-verb:"specification"
+  Spec FooSpec
+  ...
+```
+
+On CRDs schemas:
+
+```
+"x-kubernetes-permission-verb": "specification"
+```
+
+The authz system will see the verb "granular:specfication".
+
+This illustrates that the granular permissions can be made hirearchical;
+granting "granular" and "granular:specification" to some actor permits it to
+modify anything in .spec. API Server will check the permissions from most
+general to most specific, so that a single general permission can let it avoid
+checking many specific permissions.
+
+#### Pod's nodeName
+
+Given where this appears in Pod, this permission will only be checked for pods
+if the agent doesn't have the "specification" permission.
+
+This must be set on the field declaration, since nodeName is a string and not a
+struct:
+
+```go
+  ...
+  // +permission-verb:"nodeAssignment"
+  NodeName string
+  ...
+```
+
+On CRDs schemas (pod is not a CRD, but if a CRD has a field that is the same
+concept, you could add to that field this):
+
+```
+"x-kubernetes-permission-verb": "nodeAssignment"
+```
+
+The authz system will see the verb "granular:nodeAssignment".
+
+##### Everything in metadata
+
+Configured the same as everything in spec, but the verb will be (CHOICE NEEDED)
+"metadata" or "objectmeta".
+
+#### All Labels
+
+This permission will only be checked if the agent doesn't have the everything in
+metadata permission.
+
+This is set on the field declaration, since it is a map and we are granting
+permission on any item in the map.
+
+```go
+  ...
+  // +permission-verb:"labels"
+  Labels map[string]string
+  ...
+```
+
+The authz system will see the verb "granular:labels".
+
+#### Specific Labels
```

Comment:
@apelisse can you look at this section and see if you can think of a better system? (you may need to read some prior sections to understand this)